### ISSUE TITLE
Bypass traffic by domains 

### ISSUE THREAD

**Issue by asaarashi** [black-desk/cgtproxy#242]
In scenarios of high traffic load, directing traffic through cgtproxy leads to increased resource costs. Can it bypass the traffic by domains to reduce resource costs? 
If cgtproxy is not able to do that for now, please point out the method to me, and then I may be able to submit a pull request.

Thank you for the great job.

**Comment by black-desk** [comment]
I do not think iptables or nftables can filter package by domain.

**Comment by asaarashi** [comment]
Yes but I still wonder if we could use it like the combination
chinadns-ng+ss-tproxy

**Comment by asaarashi** [comment]
That sounds a big job. Perhaps I should have a new project to do  that.

**Comment by asaarashi** [comment]
Modify the codes to:
1 Inspect DNS requests through cgtproxy to get requested domains and
returned IPs
2 If the domain meets the rule specified, add a bypass rule against the IPs
to nftables

Do you think this is feasible?

**Comment by black-desk** [comment]
nft seems support dynamic ipset. You might can let the injected dns server update result ip into nft rules.

**Comment by black-desk** [comment]
You can start a new project or fork my repository.

**Comment by asaarashi** [comment]
Hi Black Desk, might I know the "injected DNS server" you mentioned means
the same as point 1 I talked about? Or an independent DNS server? Or
whatever?

**Comment by black-desk** [comment]
> Hi Black Desk, might I know the "injected DNS server" you mentioned means the same as point 1 I talked about? Or an independent DNS server I create? Or whatever?

https://github.com/black-desk/cgtproxy/blob/master/misc/config/example.yaml#L31

cgtproxy has a feature that redirect dns traffic, but cgtproxy do not have a builtin dns server.

I suggest you fork my project and add a builtin dns server which update a nft ipset according to dns response.

**Comment by asaarashi** [comment]
Yes, that's a valid approach.
However, if I still redirect the DNS requests to the proxy, I can inspect
the DNS responses instead of building a built-in DNS server, correct?

**Comment by black-desk** [comment]
> Yes, that's a valid approach. However, if I still redirect the DNS requests to the proxy, I can inspect the DNS responses instead of building a built-in DNS server, correct?

Sure but I have no idea how to let cgtproxy get the dns response without consume it, you can do some research first.

**Comment by asaarashi** [comment]
You're right. That is too difficult. Instead, I may create a built-in DNS
to intercept and forward the DNS requests to the proxy and then get the
response.

**Comment by asaarashi** [comment]
Thanks a bunch, Black Desk.
Will let you know as soon as it's successful.

**Comment by black-desk** [comment]
If it's not a problem for you, I'll close this issue as “not planned”, and if you implement the feature in your fork, we'll talk about whether or not we should merge it.

We may need to consider whether this is taking the project away from what it was designed for.

Determining whether network traffic should go through a proxy by domain name is a feature that has typically been implemented by the tproxy service. Re-implementing it in this project brings no new features.

The core feature provided by this project is the handling of network traffic by cgroup. This feature is only relevant for handling network traffic generated locally. So I wrote the code assuming that the tproxy service as well as the project are running on the endpoint that generates the network, and I don't think that the extra copy is a performance bottleneck that needs to be focused on for the web proxy service that is running locally.

Therefore, I don't think there is much practical significance in implementing such a feature.

**Comment by asaarashi** [comment]
Do you think if dns-hijack is useful in my case?

**Comment by asaarashi** [comment]
won't be a feature

**Comment by black-desk** [comment]
> Do you think if dns-hijack is helpful in my case?

If you just add a builtin dns server to cgtproxy, you can use dns-hijack feature to redirect dns request to the new dns server you add, I think.
